### ISSUE TITLE
hyperlinking of link annotations

### ISSUE THREAD

**Issue by SusanBrown** [cwrc/CWRC-Writer#369]
Hyperlinks (a link annotation) in the text do not allow you to click on the link and go to it. They should.
For links to a cwrc page, the link should load in the same window.
A new window would be preferable, I think, for links that are not to a cwrc page.

**Comment by ilovan** [comment]
As per conversation with James last Friday:
Hovering the mouse over the link/entity reference (or clicking it once on touch screens) should result in a popup that would show the page. Clicking on the link/entity ref (or clicking on it for the second time for touch screen) should open the URL/URI in another tab.

**Comment by ajmacdonald** [comment]
Functionality is in place now

**Comment by SusanBrown** [comment]
Hurray! Is there somewhere we can look at it? or do we need to pull the current code?

**Comment by ajmacdonald** [comment]
Commits to the development branch should always auto-update the dev version: http://apps.testing.cwrc.ca/editor/dev/editor_dev.htm#templates/letter

**Comment by SusanBrown** [comment]
Doesn’t seem to be working for me, either to click on the annotated strings in the text view or the entity items in the entity panel.
I cleared by cache and tried again.

**Comment by ajmacdonald** [comment]
Not sure why it's not working for you. Here's what it looks like when you mouse over a link entity:
![untitled-1](https://cloud.githubusercontent.com/assets/882444/15435302/75c5d6d6-1e89-11e6-9cc4-c1f25d6e3ea1.png)

**Comment by SusanBrown** [comment]
Aha--you're right, it definitely does this for me on link entities. But the request was meant to be to have this happen with any _linked_ entity! So in the screenshot below, it would be great to have the same effect when you mouse over Brynjulf Bull's name in the editing pane, and also if you click on the link URL in the entity panel.

![image](https://cloud.githubusercontent.com/assets/545952/15435496/7d21b30e-1e8a-11e6-8d08-8593e000b7d1.png)

**Comment by ajmacdonald** [comment]
Please re-test.

**Comment by SusanBrown** [comment]
I'm reviving this item because this doesn't seem to be working in the embedded version at least.

In this document, the words "Order this text" have a xref tag on them:

http://cwrc-dev-06.srv.ualberta.ca/islandora/object/cwrc%3A265de51b-c8de-48c4-b6e6-d4afe8f60f24/edit-cwrc#cwrc:265de51b-c8de-48c4-b6e6-d4afe8f60f24

It is not distinguished visually nor does the popup occur for the link.

The links were created in Oxygen rather than CWRC-Writer. I can't seem to attach the file but I will send it to Andrew.

I can't test how this would work in the freestanding version but I did try inserting links in one of the sample docs and the docs wouldn't validate when I did, so I'm not sure what's going on since xrefs should have been valid at that point.

**Comment by SusanBrown** [comment]
Also Andrew could you give this priority, please, over the matters I listed earlier today?

**Comment by ajmacdonald** [comment]
When a document is loaded, entity tags only get converted if the RDF in the header refers to them, or if they have an annotationId attribute. In your case the XREF tag has neither.
There exists a Convert to Entity option in the context menus which can be used on an individual basis.

When a document is loaded, it could be scanned for potential entities and the conversion done automatically. Let me know if you're interested in this functionality and how you'd expect it to work in terms of the UI: dialogs, notifications, etc.

**Comment by SusanBrown** [comment]
Re: entity tags linking. Yes it works.

I don't know what happened. I did create an entity and test whether it hyperlinked and it didn't. I did so just now again in the standalone version. It didn't "take" the first time, though the popup appeared and I filled in the dialog correctly, but then it did work find 3x in a row and the entities have appeared and are clickable in the entity panel.

Previously, though, I thought we had hyperlinking working also from the reading pane, so that if you hovered over a link entity the url would appear as hover text and you could click on it. Am I misremembering? I think it's pretty important to have links clickable from the reading pane.
___________________

Re: automatically converting entities. It would be great to have this functionality. It could be automated for xrefs, but as you say for other entities it would require interaction with the lookups. I am going to make this a separate github issue.
___________________

Re: xrefs that are not entities:

So can the contents of a simple xref tag not appear as hyperlinked text? I thought they did. We would want this. 

For instance, this item in CWRC, if you can pull it up (you should be able to, but let me know if not) contains many xref tags in the list of Archival Holdings and Published Resources:
http://beta.cwrc.ca/islandora/object/islandora%3Ae74c0d97-b7f3-4707-b826-afc11581351d#islandora:e74c0d97-b7f3-4707-b826-afc11581351d

However, none of them are showing up at all. Again, this is the embedded CWRC-Writer.
![image](https://cloud.githubusercontent.com/assets/545952/17378917/f4506874-598d-11e6-9d7c-652b4610ec01.png)

I also uploaded it in the standalone CWRC-Writer (saved as EPaulineJohnsonTPATT) by pasting it in the edit source box, and no hyperlinking worked there either.

**Comment by ajmacdonald** [comment]
"Previously, though, I thought we had hyperlinking working also from the reading pane, so that if you hovered over a link entity the url would appear as hover text and you could click on it. Am I misremembering"
This is still working in the standalone version.

xref tags can show up as links but currently they need to be identified as entities first. This is why I suggested the possibility of scanning a document for potential entities on document load.

**Comment by susanirenebrown** [comment]
While the conversion would be great, my concern is that if someone opts for the XML only option, they wouldn’t be able to get hyperlinking to work at all.

Can we not get hyperlinking working for xrefs?

**Comment by ajmacdonald** [comment]
Entities still exist in XML only mode, they just lack the RDF component.

Hyperlinks for xref tags are possible (and already get created when the user adds those tags to an already loaded document). What you're talking about is having them identified/created when the document is loaded. This is also what I'm talking about with the "conversion to entities" thing.

**Comment by SusanBrown** [comment]
Roger. So if converting them doesn’t mean creating new RDF (if they are in XML only mode) or doing making any changes to the XML markup, then let’s go the conversion route, if that seems the simplest one.

On Aug 3, 2016, at 5:51 PM, Andrew <notifications@github.com<mailto:notifications@github.com>> wrote:


Entities still exist in XML only mode, they just lack the RDF component.

Hyperlinks for xref tags are possible (and already get created when the user adds those tags to an already loaded document). What you're talking about is having them identified/created when the document is loaded. This is also what I'm talking about with the "conversion to entities" thing.

**Comment by jchartrand** [comment]
How hard will it be to implement the document scan/conversion to entities?  I’m pretty sure we’d talked about it before, and possibly decided it might be difficult?

**Comment by susanirenebrown** [comment]
What if it were XREF tags only? Everything else would be pretty complicated I think, but are these not totally straightforward conversions?

**Comment by ajmacdonald** [comment]
It was pretty simple to implement.
I just tested the extra time required to convert all the entities in the Hope McIntyre document: 11 seconds. Converting just the link entities took less than a second.
How should I proceed?

**Comment by SusanBrown** [comment]
Amazing. So ALL the entities? What did you do about the lookups?

On Aug 4, 2016, at 4:24 PM, Andrew <notifications@github.com<mailto:notifications@github.com>> wrote:


It was pretty simple to implement.
I just tested the extra time required to convert all the entities in the Hope McIntyre document: 11 seconds. Converting just the link entities took less than a second.
How should I proceed?

**Comment by ajmacdonald** [comment]
At this point all I'm doing is searching the document for tags that match the entity markup from the schema, and then using whatever info is contained in the attributes and child tags. No lookups are performed.

**Comment by susanirenebrown** [comment]
I don’t think then that there is much added value in this unless it would speed up the lookups and it sounds like it would be the same as applying a new entity tag. So unless I’m wrong about that, then I think it would be better to convert only the Xrefs.

**Comment by ajmacdonald** [comment]
It would not speed up lookups but it would populate the entities list so could easily see where entities exist in the document.

**Comment by SusanBrown** [comment]
Could we easily add an “Add Link” button to the entity instance, in the entity panel, that would spawn the lookup? That would be enough added value to make it worthwhile. 

[The other thing that would hugely add value would be the ability to drag one entity onto another to merge them--not the tags but the targets of the annotations, so that one need add the link only once. However, this would I'm sure be a major piece of work so I'm just going to break this out into another issue as a placeholder.  Plus it replicates functionality that we're developing in the named entity vetting environment that Sharcnet is developing for us.]

So, if we can add an “Add Link” button, let’s convert all tags to entities (after asking the user if they want to as follows: "Do you want to create RDF annotations from existing XML entity tags? _More info_"

My other concern is with creating a bunch of annotation that essentially have no body, or at best a literal string of the contents of the tag or of an attribute, if I'm understanding what is happening correctly. @jchartrand what's your feeling about this? Seems like bad practice to me. On the other hand, the xrefs have already all the info needed to create a decent RDF annotation.

So I'm leaning towards converting the xrefs.

**Comment by ajmacdonald** [comment]
In the entities list/panel, right clicking and selecting Edit Entity will take the user to the lookup dialog.

**Comment by SusanBrown** [comment]
Of course, so that's pretty easy and it would help people see what entities are missing links.

What goes in the body of the annotation in the absence of a URL, @ajmacdonald ? the standard or reg attribute value, or the string that's inside the tag, or what? Ideally it should be the attribute value and where one doesn't exist then the string.

@jchartrand what's your view about creating a bunch of essentially incomplete (until someone adds a URL for the entity) or at least rather useless triples?

**Comment by ajmacdonald** [comment]
Here's the RDF that currently gets produced for an XREF tag with no attributes:
```
<rdf:Description xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#" rdf:about="http://id.cwrc.ca/annotation/50b39673-0a0a-4318-8321-fa94fb51b464">
	<oa:hasTarget xmlns:oa="http://www.w3.org/ns/oa#" rdf:resource="http://id.cwrc.ca/target/f5838eb4-c698-4007-9ca2-c1c4d899b5dd"/>
	<oa:hasBody xmlns:oa="http://www.w3.org/ns/oa#" rdf:resource="http://id.cwrc.ca/link/07d04771-629c-41a5-a0a8-b4640db89ade"/>
	<oa:annotatedBy xmlns:oa="http://www.w3.org/ns/oa#" rdf:resource="http://id.cwrc.ca/user/54a470ca-532e-4916-bc07-8952e62fe41f"/>
	<oa:annotatedAt xmlns:oa="http://www.w3.org/ns/oa#">2016-08-08T16:48:21.137Z</oa:annotatedAt>
	<oa:serializedBy xmlns:oa="http://www.w3.org/ns/oa#" rdf:resource=""/>
	<oa:serializedAt xmlns:oa="http://www.w3.org/ns/oa#">2016-08-08T16:48:21.137Z</oa:serializedAt>
	<rdf:type rdf:resource="http://www.w3.org/ns/oa#Annotation"/>
	<oa:motivatedBy xmlns:oa="http://www.w3.org/ns/oa#" rdf:resource="http://www.w3.org/ns/oa#linking"/>
	<cw:cwrcAttributes xmlns:cw="http://cwrc.ca/ns/cw#">{}</cw:cwrcAttributes>
</rdf:Description>
<rdf:Description xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#" rdf:about="http://id.cwrc.ca/link/07d04771-629c-41a5-a0a8-b4640db89ade">
	<rdf:type rdf:resource="http://www.w3.org/ns/oa#SemanticTag"/>
	<rdf:type rdf:resource="http://www.w3.org/2011/content#ContentAsText"/>
</rdf:Description>
<rdf:Description xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#" rdf:about="http://id.cwrc.ca/target/f5838eb4-c698-4007-9ca2-c1c4d899b5dd">
	<oa:hasSource xmlns:oa="http://www.w3.org/ns/oa#" rdf:resource="http://id.cwrc.ca/doc/d02a3104-1991-400d-9bc6-bbe0efcc5bd1"/>
	<rdf:type rdf:resource="http://www.w3.org/ns/oa#SpecificResource"/>
	<oa:hasSelector xmlns:oa="http://www.w3.org/ns/oa#" rdf:resource="http://id.cwrc.ca/selector/48525025-ee35-4076-ad94-de44571dab70"/>
</rdf:Description>
<rdf:Description xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#" rdf:about="http://id.cwrc.ca/selector/48525025-ee35-4076-ad94-de44571dab70">
	<rdf:value>xpointer(//XREF[@annotationId="ent_87"])</rdf:value>
	<rdf:type rdf:resource="http://www.w3.org/ns/oa#FragmentSelector"/>
</rdf:Description>
```

**Comment by ajmacdonald** [comment]
Please test

**Comment by SusanBrown** [comment]
@ajmacdonald Can you please indicate more exactly what should be tested? 

Existing xrefs in documents within embedded cwrc-writer are still not showing links.

Opening within the embedded CWRC-Writer a document containing xrefs and other tags equivalent to CWRC entities does not result in the creation of new entities.

This is true when open a properly loaded text in the embedded CWRC-Writer and when I paste the same text to "load" it into the standalone.

**Comment by SusanBrown** [comment]
@ajmacdonald Can you please test on your end? Testing on our end doesn't indicate that this is fixed.

**Comment by ajmacdonald** [comment]
It looks like I'd turned off the auto-conversion of link entities at some point. I will turn this back on.

**Comment by ajmacdonald** [comment]
9cf148c06504d2dd5903a8b9407950fc41801591

**Comment by SusanBrown** [comment]
Thanks; when back on please notify

**Comment by ajmacdonald** [comment]
I've committed the code. I can't do the update in embedded versions though.

**Comment by jefferya** [comment]
@SusanBrown 

I've added the update to cwrc-dev-06 for testing. It's already on cwrc-dev-05. Let me know if want beta updated too.

**Comment by SusanBrown** [comment]
@cailinr Could you please test this? I think you should now be testing first on http://cwrc-dev-06.srv.ualberta.ca/newrepo/ but hoping for confirmation or refutation from @jefferya 

The cwrc-writer should now be taking xref tags for hyperlinks and converting them to entities upon opening the doc so that they allow users to click on the links. Playwrights Project docs have links to allow people to buy the plays associated with all their mentions of published plays.

**Comment by cailinr** [comment]
Tested in http://cwrc-dev-06.srv.ualberta.ca/newrepo/  using http://cwrc-dev-06.srv.ualberta.ca/newrepo/islandora/object/cwrc%3A265de51b-c8de-48c4-b6e6-d4afe8f60f24#cwrc:265de51b-c8de-48c4-b6e6-d4afe8f60f24, and there was no "edit" option for cwrc-writer. 

Tested in dev-05 using http://cwrc-dev-05.srv.ualberta.ca/islandora/object/cwrc%3Abcfc0a8f-4f31-485c-9a10-449e511e63de/edit-cwrc#cwrc:bcfc0a8f-4f31-485c-9a10-449e511e63de and http://cwrc-dev-05.srv.ualberta.ca/islandora/object/cwrc%3A2681e7c1-5502-43b3-bfd8-a660fba67704/edit-cwrc#cwrc:2681e7c1-5502-43b3-bfd8-a660fba67704, and there was an error loading the schema.

Tested in standalone using http://apps.testing.cwrc.ca/editor/test/editor_dev.htm#templates/Sample_Canadian_Women_Playwrights_entry and http://apps.testing.cwrc.ca/editor/test/editor_dev.htm#templates/sample_writing by adding entities. Hovering over the entities does not result in a clickable window. 

@SusanBrown

**Comment by SusanBrown** [comment]
@ajmacdonald This fell if the radar somehow and we'd like to revive it and give it high priority. Can you tackle this if you have a gap in the integration work, please, and put it next in the queue after that?

**Comment by ajmacdonald** [comment]
This has been in place for awhile. It didn't work with the Sample Canadian Women Playwrights document because that document did not contain any link entity tags (e.g. XREF), and we're only auto-converting link entity tags.
